### PR TITLE
JAMES-2341 Refactor SpamAssassinInvoker

### PR DESCRIPTION
**Pull request by aduprat**
Rebase of #1301

### CODE REVIEW

**Comment by aduprat** [comment]
test this please

File: server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinExtension.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            train(Paths.get(ClassLoader.getSystemResource("spamassassin_db/ham").toURI()), TrainingKind.HAM);
+        }
+
+        private void train(Path folder, TrainingKind trainingKind) throws URISyntaxException, IOException {
+            try (Stream<Path> paths = Files.walk(folder)) {
+                paths
+                    .filter(Files::isRegularFile)
+                    .map(Path::toFile)
+                    .map(Throwing.function(FileInputStream::new))
+                    .map(readAsPair())
```
Comment (mbaechler) [inline server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinExtension.java]: you should merge readAsPair, closeStream and Pair::getRight in a single method to make code simpler

File: server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinInvokerTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    public void scanMailShouldModifyHeadersField() throws Exception {
+        MimeMessage mimeMessage = MimeMessageUtil.mimeMessageFromStream(
+                ClassLoader.getSystemResourceAsStream("eml/spam.eml"));
+        testee.scanMail(mimeMessage);
+
+        assertThat(testee.getHeadersAsAttribute()).isNotEmpty();
+    }
+
+    @Test
+    public void scanMailShouldMarkHasSpamWhenKnownHasSpam() throws Exception {
```
Comment (rouazana) [inline server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinInvokerTest.java]: s/Has/As/g

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java [hunk 3]
```diff
@@ -78,96 +76,60 @@ public SpamAssassinInvoker(String spamdHost, int spamdPort) {
...
+    private SpamAssassinResult processSpam(String line) {
+        List<String> elements = Lists.newArrayList(Splitter.on(' ').split(line));
+        boolean spam = spam(elements.get(SPAM_INDEX));
+        String hits = elements.get(HITS_INDEX);
+        String required = elements.get(REQUIRED_HITS_INDEX);
+        SpamAssassinResult.Builder builder = SpamAssassinResult.builder()
+            .hits(hits)
+            .requiredHits(required);
+
+        if (spam) {
```
Comment (chibenwa) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: Couldn't this be a method of the builder?

```
return SpamAssassinResult.builder()
            .hits(hits)
            .requiredHits(required)
            .markAsSpam(spam)
            .build();
```

Seems more readable IMO....

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java [hunk 4]
```diff
@@ -78,96 +76,60 @@ public SpamAssassinInvoker(String spamdHost, int spamdPort) {
...
-     * 
-     * @return required The required hits before a message is handled as spam
-     */
-    public String getRequiredHits() {
-        return required;
+    private boolean spam(String string) {
+        try {
+            return Boolean.valueOf(string);
+        } catch (Exception e) {
+            return false;
```
Comment (chibenwa) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: Not even a debug log? Isn't it dangerous?

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinResult.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        public SpamAssassinResult build() {
+            Preconditions.checkNotNull(hits);
+            Preconditions.checkNotNull(requiredHits);
+            return new SpamAssassinResult(hits, requiredHits, headersAsAttribute.build());
+        }
+    }
+
+    private final String hits;
+    private final String requiredHits;
+    private final Map<String, String> headersAsAttribute;
```
Comment (chibenwa) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinResult.java]: Couldn't immutability be enforced and explicit here?

File: server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinResultTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.Test;
+
+public class SpamAssassinResultTest {
+
+    @Test
+    public void buildShouldThrowWhenHitsIsNotGiven() {
+        assertThatThrownBy(() -> SpamAssassinResult.builder()
+                .build())
+            .isInstanceOf(NullPointerException.class);
+
```
Comment (chibenwa) [inline server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinResultTest.java]: Extra line

File: server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinResultTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        
+    }
+
+    @Test
+    public void buildShouldThrowWhenRequiredHitsIsNotGiven() {
+        assertThatThrownBy(() -> SpamAssassinResult.builder()
+                .hits("1.0")
+                .build())
+            .isInstanceOf(NullPointerException.class);
+
```
Comment (chibenwa) [inline server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinResultTest.java]: Extra line

File: server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinResultTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        String name2 = "header2";
+        String value2 = "value2";
+        SpamAssassinResult spamAssassinResult = SpamAssassinResult.builder()
+            .hits(hits)
+            .requiredHits(requiredHits)
+            .putHeader(name, value)
+            .putHeader(name2, value2)
+            .build();
+
+        assertThat(spamAssassinResult.getHits()).isEqualTo(hits);
```
Comment (chibenwa) [inline server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinResultTest.java]: assert softly when you have several asserts please.

File: server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinExtension.java (hunk 1 above)
Comment (aduprat) [reply server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinExtension.java]: Reworked.

**Comment by aduprat** [comment]
#1307

**Pull request closed without merging**